x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,527 @@
+            channel_id,
+            quota_id,
+        } => try_reset_channel_quota(deps, info.sender, channel_id, quota_id, env.block.time),
+    }
+}
+
+pub struct RateLimitResponse {
+    pub rate_limit: RateLimit,
+    pub used: u128,
+    pub max: u128,
+}
+
+// Q: Is an ICS 20 transfer only 1 denom at a time, or does the caller have to split into several
+// calls if its a multi-denom ICS-20 transfer
+
+/// This function checks the rate limit and, if successful, stores the updated data about the value
+/// that has been transfered through the channel.
+/// If the period for a RateLimit has ended, the Flow information is reset.
+///
+/// The channel_value is the current value of the channel as calculated by the caller. This should
+/// be the total supply of a denom
+pub fn try_transfer(
+    deps: DepsMut,
+    sender: Addr,
+    channel_id: String,
+    channel_value: u128,
+    funds: u128,
+    direction: FlowType,
+    now: Timestamp,
+) -> Result<Response, ContractError> {
+    // Only the IBCMODULE can execute transfers
+    // TODO: Should we move this to a helper method?
+    //       This may not be needed once we move this function to be under sudo.
+    //       Though it might still be worth checking that only the transfer module is calling it
+    let ibc_module = IBCMODULE.load(deps.storage)?;
+    if sender != ibc_module {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let trackers = RATE_LIMIT_TRACKERS.may_load(deps.storage, &channel_id)?;
+
+    let configured = match trackers {
+        None => false,
+        Some(ref x) if x.is_empty() => false,
+        _ => true,
+    };
+
+    if !configured {
+        // No Quota configured for the current channel. Allowing all messages.
+        // TODO: Should there be an attribute for it being allowed vs denied?
+        return Ok(Response::new()
+            .add_attribute("method", "try_transfer")
+            .add_attribute("channel_id", channel_id)
+            .add_attribute("quota", "none"));
+    }
+
+    let mut rate_limits = trackers.unwrap();
+
+    let results: Result<Vec<RateLimitResponse>, _> = rate_limits
+        .iter_mut()
+        .map(|limit| {
+            // TODO: Should we move this to more methods on ChannelFlow?
+            // e.g. new pseudocode
+            // channel.updateIfExpired(now)
+            // channel.trackSend(&direction, funds)
+            // channel.CheckRateLimit(direction.clone())?;
+            // (Or at least update for time + rename for track send. the last one is a bit of a code style nit)
+            let max = limit.quota.capacity_at(&channel_value, &direction);
+            if limit.flow.is_expired(now) {
+                limit.flow.expire(now, limit.quota.duration)
+            }
+            limit.flow.add_flow(direction.clone(), funds);
+
+            let balance = limit.flow.balance();
+            if balance > max {
+                return Err(ContractError::RateLimitExceded {
```

Comment:
question: what happens to all the other computations and mutations happened prior to this err if errored?

Reply:
if we return an error, the transaction fails and everything gets reverted